core/shared/src/main/scala/scalaz/zio/IO.scala

```diff
@@ -358,6 +369,18 @@ sealed abstract class IO[+E, +A] extends Serializable { self =>
         }
     )(_ => self)
 
+  /**
+   * Runs the specified action if this action is interrupted.
+   */
+  final def onInterrupt(cleanup: ExitResult[E, Nothing] => IO[Nothing, Unit]): IO[E, A] =
```

Comment:
I am not sure. The types suggest that an action can be interrupted, as well as failed. And I think that's actually possible. If it's failing concurrently with being interrupted, for example. Should this count as interruption or not? The intent was to interrupt, but if it were not interrupted, it would have failed anyway (albeit, perhaps, in a recoverable / catchable way—interruption is not catchable).

Reply:
Yeah, true. I'm for interruption taking precedence over normal errors.